enzyme/Enzyme/MLIR/Interfaces/Utils.cpp

```diff
@@ -0,0 +1,374 @@
+
+  // Non-equivalent allocas/global's cannot conflict with each other
+  if ((isAlloca[0] || isGlobal[0]) && (isAlloca[1] || isGlobal[1]))
+    return false;
+
+  bool isArg[2];
+  isArg[0] = isa<BlockArgument>(v1) &&
+             isa<FunctionOpInterface>(
+                 cast<BlockArgument>(v1).getOwner()->getParentOp());
+
+  isArg[1] = isa<BlockArgument>(v1) &&
+             isa<FunctionOpInterface>(
+                 cast<BlockArgument>(v1).getOwner()->getParentOp());
+
+  // Stack allocations cannot have been passed as an argument.
+  if ((isAlloca[0] && isArg[1]) || (isAlloca[1] && isArg[0]))
+    return false;
+
+  // Non captured base allocas cannot conflict with another base value.
+  if (isAlloca[0] && !isCaptured(v1))
+    return false;
+
+  if (isAlloca[1] && !isCaptured(v2))
+    return false;
+
+  return true;
+}
+
+bool mayAlias(MemoryEffects::EffectInstance &a,
+              MemoryEffects::EffectInstance &b) {
+  if (a.getResource()->getResourceID() != b.getResource()->getResourceID())
+    return false;
+  Value valA = a.getValue();
+  Value valB = b.getValue();
+
+  // unknown effects may always alias
+  if (!valA || !valB) {
+    return true;
+  }
+
+  auto valResult = oputils::mayAlias(valA, valB);
+  return valResult;
+}
+
+bool isReadOnly(Operation *op) {
+  // If the op has memory effects, try to characterize them to see if the op
+  // is trivially dead here.
+  if (auto effectInterface = dyn_cast<MemoryEffectOpInterface>(op)) {
+    // Check to see if this op either has no effects, or only reads from memory.
+    SmallVector<MemoryEffects::EffectInstance, 1> effects;
+    effectInterface.getEffects(effects);
+    if (!llvm::all_of(effects, [op](const MemoryEffects::EffectInstance &it) {
+          return isa<MemoryEffects::Read>(it.getEffect());
+        })) {
+      return false;
+    }
+  }
+
+  bool isRecursiveContainer =
+      op->hasTrait<OpTrait::HasRecursiveMemoryEffects>();
+  if (isRecursiveContainer) {
+    for (Region &region : op->getRegions()) {
+      for (auto &block : region) {
+        for (auto &nestedOp : block)
+          if (!isReadOnly(&nestedOp))
+            return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+std::optional<SmallVector<MemoryEffects::EffectInstance>>
+collectOpEffects(Operation *rootOp) {
+  SmallVector<MemoryEffects::EffectInstance> effects;
+  SmallVector<Operation *> effectingOps(1, rootOp);
+  while (!effectingOps.empty()) {
+    Operation *op = effectingOps.pop_back_val();
+    bool isRecursiveContainer =
+        op->hasTrait<OpTrait::HasRecursiveMemoryEffects>();
+    if (isRecursiveContainer) {
+      for (Region &region : op->getRegions()) {
+        for (Block &block : region) {
+          for (Operation &nestedOp : block) {
+            effectingOps.push_back(&nestedOp);
+          }
+        }
+      }
+    }
+
+    if (auto effectInterface = dyn_cast<MemoryEffectOpInterface>(op)) {
+      effectInterface.getEffects(effects);
```

Comment:
I'm not sure this is guaranteed to append into effects [hence the old code made a temporary allocation for the local effects, then appended]